Added function to create block from boundary

Feuerbowle opened the pull request:
There are two new functions
CreateBlock(PMBoundary boundary)
creates block from boundary and calculates the heights automatically

CreateBlock(PMBoundary boundary, double ZMin, double ZMax)
creates block from boundary and ZMin and ZMax values

Feuerbowle committed:
Added function to create block from boundary

There are two new functions
CreateBlock(PMBoundary boundary)
creates block from boundary and calculates the heights automatically

CreateBlock(PMBoundary boundary, double ZMin, double ZMax)
creates block from boundary and ZMin and ZMax values

lukeedw commented on Delcam.ProductInterface.PowerMILL.Test/PMEntityTests/PMProjectTests.cs:
```diff
@@ -125,6 +125,30 @@ public void ActiveWorldWorkplaneTest()
+        [Test]
+        public void CreateBlockFromBoundary()
+        {   
+            _powerMILL.Execute(string.Format("CREATE BOUNDARY \"{0}\"", "TestBoundary"));            
```
Why not use the API code to create the boundary from a file?

lukeedw commented on Delcam.ProductInterface.PowerMILL.Test/PMEntityTests/PMProjectTests.cs:
```diff
@@ -125,6 +125,30 @@ public void ActiveWorldWorkplaneTest()
+            _powerMILL.ActiveProject.Refresh();
+            PMBoundary boundary = _powerMILL.ActiveProject.Boundaries.ActiveItem;
+            _powerMILL.ActiveProject.CreateBlock(boundary, 0, 100);
+            Assert.AreEqual("100,0", _powerMILL.ExecuteEx("PRINT $BLOCK.LIMITS.ZMax"));
```
Can we add properties for the block limits while we are here?  Perhaps we can get a BoundingBox back?

Feuerbowle replied:
Do you mean something like this?
`Autodesk.Geometry.BoundingBox boundingBox = _powerMILL.ActiveProject.CreateBlockFromBoundaryWithLimits(boundary, 0, 100);
Assert.AreEqual(100, boundingBox.MaxZ);`

lukeedw replied:
Yes, looks great.  An even better way to do the assert is:
Assert.That(boundingBox.MaxZ, Is.EqualTo(100));

lukeedw commented on Delcam.ProductInterface.PowerMILL/Classes/PMProject.cs:
```diff
@@ -423,6 +423,36 @@ public void CreateBlock(string modelName, double blockExpansion)
+        /// This operation creates a Block in PowerMILL from an boundary. The heights are calculated automatically 
+        /// </summary>
+        /// <param name="boundary">The boundary from which to create the block.</param>        
+        public void CreateBlock(PMBoundary boundary)
```
Try to avoid overloading.  We have found over time that you can end up with a lot of functions with the same name and it isn't clear what they are doing different to each other.  CreateBlockFromBoundary would be good here.  CreateBlockFromBoundaryWithLimits for the other.

Feuerbowle committed:
Changes made as requested

Now using the API to create the boundary from file.
No function overload.
The functions now return a boundingBox with the limits of the block.

lukeedw commented on Delcam.ProductInterface.PowerMILL.Test/PMEntityTests/PMProjectTests.cs:
```diff
@@ -126,27 +126,25 @@ public void ActiveWorldWorkplaneTest()
-            Assert.AreEqual("-30,004846", _powerMILL.ExecuteEx("PRINT $BLOCK.LIMITS.ZMin"));            
+            PMBoundary boundary = _powerMILL.ActiveProject.Boundaries.ActiveItem;
+            Autodesk.Geometry.BoundingBox boundingBox = _powerMILL.ActiveProject.CreateBlockFromBoundary(boundary);
+            Assert.That(Math.Round(boundingBox.MaxZ.Value, 5), Is.EqualTo(53.35378));
```
Nicer way to write this is:
Assert.That(boundingBox.MaxZ, Is.EqualTo((MM)53.35378));
The comparison for MM contains a rounding check so it should just work...

Feuerbowle replied:
Nice to know. Thanks

lukeedw commented on Delcam.ProductInterface.PowerMILL.Test/PMEntityTests/PMProjectTests.cs:
```diff
@@ -126,27 +126,25 @@ public void ActiveWorldWorkplaneTest()
             PMBoundary boundary = _powerMILL.ActiveProject.Boundaries.ActiveItem;
-            _powerMILL.ActiveProject.CreateBlock(boundary, 0, 100);
-            Assert.AreEqual("100,0", _powerMILL.ExecuteEx("PRINT $BLOCK.LIMITS.ZMax"));
+            Autodesk.Geometry.BoundingBox boundingBox = _powerMILL.ActiveProject.CreateBlockFromBoundaryWithLimits(boundary, 0, 100);            
```
One nice thing with C# is that rather than having to specify the full type in the declaration you can use "var" instead.  So write:
var boundingBox = _powerMILL.ActiveProject.CreateBlockFromBoundaryWithLimits(....)
and let the compiler do the hard work!

lukeedw commented on Delcam.ProductInterface.PowerMILL/Classes/PMProject.cs:
```diff
@@ -451,6 +454,27 @@ public void CreateBlock(PMBoundary boundary, double ZMin, double ZMax)
+        /// </summary>
+        public BoundingBox GetBlockLimits()
+        {
+            string tpBlockXmin = _powerMILL.DoCommandEx("PRINT PAR $Block.Limits.XMin").ToString().Replace("(REAL)", "").Replace("[L]", "").Trim();
```
If you use "PRINT PAR TERSE" rather than "PRINT PAR" then you don't need to get rid of the "(REAL)" string.

Feuerbowle replied:
I've always wondered what "TERSE" do

Feuerbowle committed:
Further changes

Changed the Assert.That in the CreateBlockFromBoundaryTest() function
Declaration of the boundingBox changed to var
"PRINT PAR" changed to "PRINT PAR TERSE"
